ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/CommandRunnerTest.java

```diff
@@ -167,6 +182,75 @@ public void shouldEarlyOutIfNewCommandsContainsTerminate() {
     verify(statementExecutor, never()).handleRestore(queuedCommand3);
   }
 
+  @Test
+  public void shouldReportRunningIfNotStuckProcessingCommand() throws BrokenBarrierException, InterruptedException, ExecutionException {
+    try {
+      checkCommandRunnerStatus(
+          COMMAND_RUNNER_HEALTH_TIMEOUT - 500,
+          COMMAND_RUNNER_HEALTH_TIMEOUT - 1000,
+          CommandRunner.CommandRunnerStatus.RUNNING
+      );
+    } catch (Exception e) {
+      // fail test if an exception happens
+      assertThat(true, equalTo(false));
+    }
+  }
+
+  @Test
+  public void shouldReportErrorIfStuckProcessingCommand() throws BrokenBarrierException, InterruptedException, ExecutionException {
+    try {
+      checkCommandRunnerStatus(
+          COMMAND_RUNNER_HEALTH_TIMEOUT + 1000,
+          COMMAND_RUNNER_HEALTH_TIMEOUT + 500,
+          CommandRunner.CommandRunnerStatus.ERROR
+      );
+    } catch (Exception e) {
+      // fail test if an exception happens
+      assertThat(true, equalTo(false));
+    }
+  }
+
+  private void checkCommandRunnerStatus(
```

Comment:
I think the spirit of this test is great - if it passes we're confident the code under test is doing what its supposed to do. The problem is that because it relies on timing, it's prone to spurious test failures. I think we can make a couple tweaks to make the test deterministic:
    - pass a mock clock to command runner so we control the time changes
    - instead of using a sleep to simulate delays, have the command runner wait on a condition or countdown latch.

so you get something like this:

```
...
givenQueuedCommands(queuedCommand1);
Producer<Long> clock = mock(Producer.class);
CountDownLatch latch = new CountDownLatch(1);
CommandRunner commandRunner = new CommandRunner(..., clock::get, ...);
when(clock.get()).thenReturn(0).thenReturn(500).thenReturn(1000).thenReturn(2000);
when(statementExecutor.handleStatement()).thenAnswer(i -> latch.await());
Thread t = new Thread(() -> {
    commandRunner.fetchAndRunCommands();
});
assertThat(commandRunner.checkCommandRunnerStatus(), is(RUNNING));
assertThat(commandRunner.checkCommandRunnerStatus(), is(ERROR));
latch.countDown();
t.join();
assertThat(commandRunner.checkCommandRunnerStatus(), is(RUNNING));
...
```

Reply:
updated